src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Hosting;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.WebApi
+{
+    internal sealed class EgressOperationService : BackgroundService
+    {
+        private EgressOperationQueue _queue;
+        private IServiceProvider _serviceProvider;
+        private EgressOperationStore _operationStore;
+
+        public EgressOperationService(EgressOperationQueue taskQueue,
+            IServiceProvider serviceProvider,
+            EgressOperationStore operationStore)
+        {
+            _queue = taskQueue;
+            _serviceProvider = serviceProvider;
+            _operationStore = operationStore;
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                var item = await _queue.DequeueAsync(stoppingToken);
+
+                //We have two stopping tokens, one per item that can be triggered via Delete
+                //and if we are stopping the service
+                using (CancellationTokenSource linkedTokenSource =
+                    CancellationTokenSource.CreateLinkedTokenSource(item.CancellationTokenSource.Token, stoppingToken))
+                {
+                    CancellationToken token = linkedTokenSource.Token;
+                    var result = await item.EgressOperation.ExecuteAsync(_serviceProvider, token);
```

Comment:
This approach will not work; it forces serialization of all requests, and indefinite requests will block all other requests from the queue.

Reply:
Maybe you can create another async method that effectively does a Task.WhenAny on the in-flight executions + one task from a completion source that is signaled when when a new operation is dequeued. If that last task is signaled, just loop back around and start a new Task.WhenAny. If any other task, complete the operation.

Another option is to add continuations to each of the tasks that completes the operation when it finishes.